src/jalv.c

```diff
@@ -1064,24 +1050,66 @@ jalv_open(Jalv* const jalv, int argc, char** argv)
 		  sizeof(float), jalv->urids.atom_Float, &jalv->ui_update_hz },
 		{ LV2_OPTIONS_INSTANCE, 0, 0, 0, 0, NULL }
 	};
+	memcpy(jalv->features.options, options, sizeof(jalv->features.options));
 
-	options_feature.data = (void*)&options;
+	init_feature(&jalv->features.options_feature,
+	             LV2_OPTIONS__options,
+	             (void*)jalv->features.options);
+
+	init_feature(&jalv->features.safe_restore_feature,
+	             LV2_STATE__threadSafeRestore,
+	             NULL);
 
 	/* Create Plugin <=> UI communication buffers */
 	jalv->ui_events     = zix_ring_new(jalv->opts.buffer_size);
 	jalv->plugin_events = zix_ring_new(jalv->opts.buffer_size);
 	zix_ring_mlock(jalv->ui_events);
 	zix_ring_mlock(jalv->plugin_events);
 
+	/* Build feature list for passing to plugins */
+	const LV2_Feature* const features[] = {
+		&jalv->features.map_feature,
+		&jalv->features.unmap_feature,
+		&jalv->features.sched_feature,
+		&jalv->features.log_feature,
+		&jalv->features.options_feature,
+		&static_features[0],
+		&static_features[1],
+		&static_features[2],
+		&static_features[3],
+		&static_features[4],
+		NULL
+	};
+	jalv->feature_list = calloc(1, sizeof(features));
+	if (!jalv->feature_list) {
+		fprintf(stderr, "Failed to allocate feature list\n");
+		jalv_close(jalv);
+		return EXIT_FAILURE;
```

Comment:
@drobilla 
I am not sure if it is guaranteed that `EXIT_FAILURE =! 0` always on all platforms. If not the error check in main() would fail.
I would replace it with a hard-coded negative value. What do you think?

Reply:
`0` is success and non-zero is non success on any platform that even remotely matters.  Negative doesn't mean anything in this context, but I have no problem with returning an explicit number here.